### PR TITLE
Fix tot magnetization

### PR DESCRIPTION
**Pull request by AndresOrtegaGuerrero**
This PR solves #412

### CODE REVIEW

**Comment by AndresOrtegaGuerrero** [comment]
@unkcpz  This is a draft, but in order for me to continue there must be a PR (in aiida-quantumespresso) to deal with the logic that if the system has` tot_magnetization` , it should remove `starting_magnetization `for PwBaseWorkChain. Likewise that the logic in PdosWorkChain and BandsWorkchain is not affected

File: src/aiidalab_qe/app/configuration/advanced.py [hunk 1]
```diff
@@ -176,6 +176,11 @@ def _update_input_structure(self, change):
             self.pseudo_setter.structure = change["new"]
             self._display_mesh()
 
+    @tl.observe("electronic_type")
+    def _electronic_type_changed(self, change):
+        """Input electronic_type changed, update the widget values."""
+        self.magnetization.electronic_type = change["new"]
```
Comment (unkcpz) [inline src/aiidalab_qe/app/configuration/advanced.py:182]: I like this, very clear to me and I saw you made `electronic_type` a trait of `MagnetizationSettings`, exactly what I think about ;)

File: src/aiidalab_qe/app/configuration/advanced.py [hunk 2]
```diff
@@ -237,12 +242,10 @@ def get_panel_value(self):
...
         parameters["pw"]["parameters"]["SYSTEM"]["tot_charge"] = self.total_charge.value
-        # there are two choose, use link or parent
-        if self.spin_type == "collinear":
-            parameters[
-                "initial_magnetic_moments"
-            ] = self.magnetization.get_magnetization()
+
+        # add the kpoints_distance to the parameters
         parameters["kpoints_distance"] = self.value.get("kpoints_distance")
+
```
Comment (unkcpz) [inline src/aiidalab_qe/app/configuration/advanced.py:248]: Can you write a comment here for all conditions? 
For instance:

```
# insulator and colliner: xxxx
# metal and collinear: xxx
```

Can the if..else clause below can be simplified with inverse and break return?

File: src/aiidalab_qe/app/configuration/advanced.py [hunk 3]
```diff
@@ -302,8 +328,11 @@ def reset(self):
             self.override.value = False
             self.smearing.reset()
             # reset the pseudo setter
-            self.pseudo_setter.structure = None
-            self.pseudo_setter._reset()
+            if self.input_structure is None:
+                self.pseudo_setter.structure = None
+                self.pseudo_setter._reset()
+            else:
+                self.pseudo_setter._reset()
```
Comment (unkcpz) [inline src/aiidalab_qe/app/configuration/advanced.py:335]: Why can not call pseudo_setter.structure = None if input_structure is not `None`? If I understand correct, the reset should also do `self.structure=None` first.

File: src/aiidalab_qe/app/configuration/advanced.py [hunk 4]
```diff
@@ -333,45 +362,67 @@ class MagnetizationSettings(ipw.VBox):
...
+        self.magnetization_type.disabled = self.disabled
 
     def reset(self):
         self.disabled = True
-        self.kinds = None
-        self.description.value = "Define magnetization: Input structure not confirmed"
-        with self.kinds_widget_out:
-            clear_output()
-            display(self.kinds)
+        self.tot_magnetization.value = 0.0
```
Comment (unkcpz) [inline src/aiidalab_qe/app/configuration/advanced.py:416]: I'd suggest making the default value defined as class attribute `_DEFAULT_TOT_MAGNETIZATION`.

File: src/aiidalab_qe/app/configuration/advanced.py [hunk 5]
```diff
@@ -333,45 +362,67 @@ class MagnetizationSettings(ipw.VBox):
...
+        #
+        if self.input_structure is None:
+            self.description.value = (
+                "<b>Magnetization: Input structure not confirmed</b>"
```
Comment (unkcpz) [inline src/aiidalab_qe/app/configuration/advanced.py:420]: This also appears in the `__init__`, which should be a way to only define it once in a single place.

File: src/aiidalab_qe/app/configuration/advanced.py [hunk 6]
```diff
@@ -392,11 +443,30 @@ def create_kinds_widget(self):
...
+                    clear_output()
+
     def update_kinds_widget(self):
         self.input_structure_labels = self.input_structure.get_kind_names()
         self.kinds = self.create_kinds_widget()
-        self.description.value = "Define magnetization: "
-        self.display_kinds()
+        self.description.value = "<b>Magnetization</b>"
+
+    def _display_starting(self, value):
```
Comment (unkcpz) [inline src/aiidalab_qe/app/configuration/advanced.py:463]:
```suggestion
    def _render(self, value):
```

File: src/aiidalab_qe/plugins/bands/workchain.py [hunk 7]
```diff
@@ -211,6 +211,11 @@ def get_builder(codes, structure, parameters, **kwargs):
     bands.pop("relax")
     bands.pop("structure", None)
     bands.pop("clean_workdir", None)
+
+    if scf_overrides["pw"]["parameters"]["SYSTEM"].get("tot_magnetization") is not None:
+        bands.scf["pw"]["parameters"]["SYSTEM"].pop("starting_magnetization", None)
+        bands.bands["pw"]["parameters"]["SYSTEM"].pop("starting_magnetization", None)
```
Comment (unkcpz) [inline src/aiidalab_qe/plugins/bands/workchain.py:217]: If I understand correctly, the code here is for when `tot_magnetization` is set, the `starting_magnetization` should not be set, but this exclusive existence can be guaranteed when the builder is created, no?

**Review by unkcpz — CHANGES_REQUESTED** [review]
Thanks a lot for trying this @AndresOrtegaGuerrero, The logic looks all good, but I think it requires some iron-up on the clarity of implementation. Can you fix the conflicting and we arrange a pair coding session to check things together?

**Comment by AndresOrtegaGuerrero** [comment]
To complete this PR the following requirements should be met

- [ ] PdosWorkChain should be adapted to run calculations using tot_magnetization https://github.com/aiidateam/aiida-quantumespresso/issues/988
- [ ] the logic of the qe app should be adapted for having two fermi energies (up/down): get_pdos_data, projections, bands-widget

**Review by superstar54 — APPROVED** [review]
@AndresOrtegaGuerrero , thanks for the work!  LGTM!

There are two things we could improve:
- If users set `Magnetism` to `Off`, the Magnetization section could be hidden.
- When switching on `Magnetism`, the pdos plot could has an optios to plot `up` and `down` in two sides.
Could you create issues on this? so that we can do these in the future PRs.

**Comment by superstar54** [comment]
Before merging the PR, two small changes are needed:

- the `tot_magnetization` should be an integer for a `fixed` occupation.
- There are two `tot_magnetization` when I load the previous job.
image

**Comment by AndresOrtegaGuerrero** [comment]
@superstar54 thank you for the review, I thought i fixed the duplication i will have a look. Regarding the tot_magnetization , i kept as a Real number since in the pw.x manual is a real.

**Comment by superstar54** [comment]
In the fixed occupation, if one sets a non-integer value for the `tot_magnetization,` pw will raise an error. In the smearing case, it will not.

**Comment by AndresOrtegaGuerrero** [comment]
> In the fixed occupation, if one sets a non-integer value for the `tot_magnetization,` pw will raise an error. In the smearing case, it will not.

Then i guess it should be conditioned then , thank you !

**Pull request merged**